### ISSUE TITLE
python packaging fails without Internet access

### ISSUE THREAD

**Issue by czanik** [syslog-ng/syslog-ng#4223]
Trying to compile the latest syslog-ng git snapshot I ran into a problem: the build process tries to use the Internet. There is no Internet access in the openSUSE build service, and turned off by default in the Fedora COPR build service. FreeBSD packages are also built without Internet access. I do not follow Debian / Ubuntu closely, but AFAIR they do not allow Internet access either.

```
[  107s] if [ "venv" = "venv" ]; then \
[  107s]     ./scripts/build-python-venv.sh "/usr/bin/python3" "/home/abuild/rpmbuild/BUILD/syslog-ng-3.38.1.248.gbc48493/venv" "."; \
[  107s] fi
[  107s] Building dev virtualenv for syslog-ng at /home/abuild/rpmbuild/BUILD/syslog-ng-3.38.1.248.gbc48493/venv from ./dev-requirements.txt
[  112s] WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x56128a67a728>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',)': /simple/cachetools/
[  113s] WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x56128a6dd648>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',)': /simple/cachetools/
[  114s] WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x56128a6cc748>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',)': /simple/cachetools/
[  116s] WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x56128a5d4f48>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',)': /simple/cachetools/
[  120s] WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x56128a6e2728>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',)': /simple/cachetools/
[  120s] ERROR: Could not find a version that satisfies the requirement cachetools==4.2.4 (from -r ./requirements.txt (line 1)) (from versions: none)
[  120s] ERROR: No matching distribution found for cachetools==4.2.4 (from -r ./requirements.txt (line 1))
[  120s] make[4]: *** [Makefile:29295: /home/abuild/rpmbuild/BUILD/syslog-ng-3.38.1.248.gbc48493/.python-venv-built] Error 1
[  120s] make[3]: *** [Makefile:28203: install-exec-am] Error 2
[  120s] make[2]: *** [Makefile:26911: install-am] Error 2
[  120s] make[1]: *** [Makefile:25118: install-recursive] Error 1
[  120s] make: *** [Makefile:26904: install] Error 2
[  120s] error: Bad exit status from /var/tmp/rpm-tmp.dQhhv2 (%install)
```

**Comment by czanik** [comment]
Now, that Kubernetes support is merged, package building fails again. It would be pretty useful, if turning off building/packaging Python packages would be possible (just like with Java). For Linux packagers who do not want to enable them, or BSD packagers, where there is no Kubernetes support at all, but Python is in active use. Unfortunately, build fails with:

```
[   90s] (cd ./modules/python-modules && \
[   90s]  mkdir -p /home/abuild/rpmbuild/BUILD/syslog-ng-3.38.1.344.gd829200/modules/python-modules && \
[   90s]          /usr/bin/python3 setup.py egg_info --egg-base="/home/abuild/rpmbuild/BUILD/syslog-ng-3.38.1.344.gd829200/modules/python-modules"  && \
[   90s]          if [ -f /home/abuild/rpmbuild/BUILD/syslog-ng-3.38.1.344.gd829200/modules/python-modules/syslogng.egg-info/requires.txt ]; then /usr/bin/python3 -m pip install -r /home/abuild/rpmbuild/BUILD/syslog-ng-3.38.1.344.gd829200/modules/python-modules/syslogng.egg-info/requires.txt; fi)
[   91s] running egg_info
[   91s] creating /home/abuild/rpmbuild/BUILD/syslog-ng-3.38.1.344.gd829200/modules/python-modules/syslogng.egg-info
[   91s] writing /home/abuild/rpmbuild/BUILD/syslog-ng-3.38.1.344.gd829200/modules/python-modules/syslogng.egg-info/PKG-INFO
[   91s] writing dependency_links to /home/abuild/rpmbuild/BUILD/syslog-ng-3.38.1.344.gd829200/modules/python-modules/syslogng.egg-info/dependency_links.txt
[   91s] writing requirements to /home/abuild/rpmbuild/BUILD/syslog-ng-3.38.1.344.gd829200/modules/python-modules/syslogng.egg-info/requires.txt
[   91s] writing top-level names to /home/abuild/rpmbuild/BUILD/syslog-ng-3.38.1.344.gd829200/modules/python-modules/syslogng.egg-info/top_level.txt
[   91s] writing manifest file '/home/abuild/rpmbuild/BUILD/syslog-ng-3.38.1.344.gd829200/modules/python-modules/syslogng.egg-info/SOURCES.txt'
[   91s] reading manifest file '/home/abuild/rpmbuild/BUILD/syslog-ng-3.38.1.344.gd829200/modules/python-modules/syslogng.egg-info/SOURCES.txt'
[   91s] writing manifest file '/home/abuild/rpmbuild/BUILD/syslog-ng-3.38.1.344.gd829200/modules/python-modules/syslogng.egg-info/SOURCES.txt'
[   91s] Defaulting to user installation because normal site-packages is not writeable
[   91s] WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x557c4fff0f08>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',)': /simple/kubernetes/
[   92s] WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x557c4ffb3cb8>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',)': /simple/kubernetes/
[   93s] WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x557c500079b8>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',)': /simple/kubernetes/
[   95s] WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x557c4ff3aff8>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',)': /simple/kubernetes/
[   99s] WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x557c4ffd9408>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',)': /simple/kubernetes/
[   99s] ERROR: Could not find a version that satisfies the requirement kubernetes (from -r /home/abuild/rpmbuild/BUILD/syslog-ng-3.38.1.344.gd829200/modules/python-modules/syslogng.egg-info/requires.txt (line 1)) (from versions: none)
[   99s] ERROR: No matching distribution found for kubernetes (from -r /home/abuild/rpmbuild/BUILD/syslog-ng-3.38.1.344.gd829200/modules/python-modules/syslogng.egg-info/requires.txt (line 1))
[   99s] make[4]: *** [Makefile:29633: pymodules-install-requirements] Error 1
[   99s] make[3]: *** [Makefile:28238: install-exec-am] Error 2
[   99s] make[2]: *** [Makefile:26945: install-am] Error 2
[   99s] make[1]: *** [Makefile:25152: install-recursive] Error 1
[   99s] make: *** [Makefile:26938: install] Error 2
```

**Comment by czanik** [comment]
I did distro packaging for openSUSE / SLES. The necessary dependencies are included in Tumbleweed and the latest Leap and SLES service pack, but not in earlier, still supported versions.

**Comment by bazsi** [comment]
I have now added the requested configure option, I am not entirely sure however that this is the right way forward.

Right now, it's only about Kubernetes, but I have another Python based plugin queued and I imagine there would be more over time. The increasing number of syslog-ng functionality, implemented in Python grows, eventually we will depend on more Python libraries. How should those be managed on FreeBSD?

**Comment by czanik** [comment]
@bazsi  Once there are more Python modules, would it be possible to select what to disable? I mean ```--disable-python-modules``` would disable all Python modules, and ``--disable-python-modules=kubernetes,another``` would disable only the listed modules ("kubernetes" and "another" in this case). Of course that would require per module dependency handling, as now everything is listed in https://github.com/syslog-ng/syslog-ng/blob/master/requirements.txt as far as I can understand. But for now the current version is good enough.

**Comment by czanik** [comment]
@bazsi Just tried to compile the latest git snapshot on FreeBSD with --disable-python-modules and it seems to disable all the Python code. Does Python support work without these files?
```
===> Checking for items in pkg-plist which are not in STAGEDIR
Error: Missing: lib/syslog-ng/python/syslogng-1.0-py%%PYTHON_VER%%.egg-info/PKG-INFO
Error: Missing: lib/syslog-ng/python/syslogng-1.0-py%%PYTHON_VER%%.egg-info/SOURCES.txt
Error: Missing: lib/syslog-ng/python/syslogng-1.0-py%%PYTHON_VER%%.egg-info/dependency_links.txt
Error: Missing: lib/syslog-ng/python/syslogng-1.0-py%%PYTHON_VER%%.egg-info/top_level.txt
Error: Missing: lib/syslog-ng/python/syslogng/__init__.py
Error: Missing: lib/syslog-ng/python/syslogng/__pycache__/__init__%%PYTHON_EXT_SUFFIX%%.pyc
Error: Missing: lib/syslog-ng/python/syslogng/__pycache__/dest%%PYTHON_EXT_SUFFIX%%.pyc
Error: Missing: lib/syslog-ng/python/syslogng/__pycache__/logger%%PYTHON_EXT_SUFFIX%%.pyc
Error: Missing: lib/syslog-ng/python/syslogng/__pycache__/message%%PYTHON_EXT_SUFFIX%%.pyc
Error: Missing: lib/syslog-ng/python/syslogng/__pycache__/parser%%PYTHON_EXT_SUFFIX%%.pyc
Error: Missing: lib/syslog-ng/python/syslogng/__pycache__/persist%%PYTHON_EXT_SUFFIX%%.pyc
Error: Missing: lib/syslog-ng/python/syslogng/__pycache__/source%%PYTHON_EXT_SUFFIX%%.pyc
Error: Missing: lib/syslog-ng/python/syslogng/__pycache__/template%%PYTHON_EXT_SUFFIX%%.pyc
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/__init__.py
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/__pycache__/__init__%%PYTHON_EXT_SUFFIX%%.pyc
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/__pycache__/choicecompleter%%PYTHON_EXT_SUFFIX%%.pyc
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/__pycache__/commandlinelexer%%PYTHON_EXT_SUFFIX%%.pyc
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/__pycache__/completer%%PYTHON_EXT_SUFFIX%%.pyc
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/__pycache__/completerlang%%PYTHON_EXT_SUFFIX%%.pyc
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/__pycache__/debuggercli%%PYTHON_EXT_SUFFIX%%.pyc
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/__pycache__/debuglang%%PYTHON_EXT_SUFFIX%%.pyc
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/__pycache__/editline%%PYTHON_EXT_SUFFIX%%.pyc
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/__pycache__/getoptlexer%%PYTHON_EXT_SUFFIX%%.pyc
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/__pycache__/langcompleter%%PYTHON_EXT_SUFFIX%%.pyc
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/__pycache__/lexer%%PYTHON_EXT_SUFFIX%%.pyc
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/__pycache__/lexertoken%%PYTHON_EXT_SUFFIX%%.pyc
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/__pycache__/macrocompleter%%PYTHON_EXT_SUFFIX%%.pyc
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/__pycache__/readline%%PYTHON_EXT_SUFFIX%%.pyc
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/__pycache__/syslognginternals%%PYTHON_EXT_SUFFIX%%.pyc
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/__pycache__/tablexer%%PYTHON_EXT_SUFFIX%%.pyc
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/__pycache__/templatelang%%PYTHON_EXT_SUFFIX%%.pyc
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/__pycache__/templatelexer%%PYTHON_EXT_SUFFIX%%.pyc
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/__pycache__/tflang%%PYTHON_EXT_SUFFIX%%.pyc
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/choicecompleter.py
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/commandlinelexer.py
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/completer.py
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/completerlang.py
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/debuggercli.py
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/debuglang.py
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/editline.py
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/getoptlexer.py
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/langcompleter.py
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/lexer.py
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/lexertoken.py
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/macrocompleter.py
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/readline.py
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/syslognginternals.py
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/tablexer.py
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/templatelang.py
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/templatelexer.py
Error: Missing: lib/syslog-ng/python/syslogng/debuggercli/tflang.py
Error: Missing: lib/syslog-ng/python/syslogng/dest.py
Error: Missing: lib/syslog-ng/python/syslogng/logger.py
Error: Missing: lib/syslog-ng/python/syslogng/message.py
Error: Missing: lib/syslog-ng/python/syslogng/modules/example/__init__.py
Error: Missing: lib/syslog-ng/python/syslogng/modules/example/__pycache__/__init__%%PYTHON_EXT_SUFFIX%%.pyc
Error: Missing: lib/syslog-ng/python/syslogng/modules/example/scl/example.conf
Error: Missing: lib/syslog-ng/python/syslogng/parser.py
Error: Missing: lib/syslog-ng/python/syslogng/persist.py
Error: Missing: lib/syslog-ng/python/syslogng/source.py
Error: Missing: lib/syslog-ng/python/syslogng/template.py
Error: Missing: %%DATADIR%%/include/scl/kubernetes/kubernetes.conf
===> Error: Plist issues found.
*** Error code 1
```

**Comment by bazsi** [comment]
Pfff, the issue is that I merged the core syslog-ng code with the python
based modules on top, so neither are installed. :(

I'll have to think how this could be solved. Selective installation of
python modules is not something I would like to support on the makefile
level. It is doable at the distro packaging, by copying the associated
files verbatim to the target location.

Let me think about this as well.

**Comment by czanik** [comment]
@bazsi Well, a bit more testing revealed that the RHEL 8 package uses Python 3.9, but the Python dependencies are only available as Python 3.6 packages. So, I had to use --disable-python-modules both for RHEL 7 and RHEL 8

For distributions where all the dependencies are available as packages, I'd prefer to use system packages. But for others it would be nice if syslog-ng Python modules could be copied in place even without having the dependencies available during installation time. And then ```/usr/bin/syslog-ng-update-virtualenv``` could finish the setup on the installed system.

Control should be more granular. I mean, Kubernetes-related modules do not make sense on FreeBSD, but I'd say that it does not make sense on my desktop machine or a Mac either.
